src/daft-sql/src/functions.rs

```diff
@@ -1,635 +1,172 @@
-    Length,
-    /// SQL 'lower()' function
-    /// # Example
-    /// ```sql
-    /// SELECT lower('HELLO');
-    /// ```
-    Lower,
-    /// SQL 'upper()' function
-    /// # Example
-    /// ```sql
-    /// SELECT upper('hello');
-    /// ```
-    Upper,
-    /// SQL 'lstrip()' function
-    /// # Example
-    /// ```sql
-    /// SELECT lstrip(' hello');
-    /// ```
-    Lstrip,
-    /// SQL 'rstrip()' function
-    /// # Example
-    /// ```sql
-    /// SELECT rstrip('hello ');
-    /// ```
-    Rstrip,
-    /// SQL 'reverse()' function
-    /// # Example
-    /// ```sql
-    /// SELECT reverse('olleh');
-    /// ```
-    Reverse,
-    /// SQL 'capitalize()' function
-    /// # Example
-    /// ```sql
-    /// SELECT capitalize('hello');
-    /// ```
-    Capitalize,
-    /// SQL 'left()' function
-    /// # Example
-    /// ```sql
-    /// SELECT left('hello', 2);
-    /// ```
-    Left,
+/// Current feature-set
+fn check_features(func: &Function) -> SQLPlannerResult<()> {
+    if func.filter.is_some() {
+        // <agg>(..) FILTER (WHERE ..)
+        unsupported_sql_err!("Aggregation `FILTER`");
+    }
+    if func.over.is_some() {
+        // WINDOW functions
+        unsupported_sql_err!("Window functions `OVER`");
+    }
+    if !func.within_group.is_empty() {
+        // <agg>(...) WITHIN GROUP
+        unsupported_sql_err!("Aggregation `WITHIN GROUP`");
+    }
+    if func.null_treatment.is_some() {
+        // Snowflake NULL treatment
+        unsupported_sql_err!("Window function `IGNORE|RESPECT NULLS`");
+    }
+    if func.parameters != FunctionArguments::None {
+        // ClickHouse parametric functions
+        unsupported_sql_err!("Parameterized function calls");
+    }
+    Ok(())
+}
 
-    /// SQL 'right()' function
-    /// # Example
-    /// ```sql
-    /// SELECT right('hello', 2);
-    /// ```
-    Right,
-    /// SQL 'find()' function
-    /// # Example
-    /// ```sql
-    /// SELECT find('hello', 'l');
-    /// ```
-    Find,
-    /// SQL 'rpad()' function
-    /// # Example
-    /// ```sql
-    /// SELECT rpad('hello', 10, ' ');
-    /// ```
-    Rpad,
-    /// SQL 'lpad()' function
-    /// # Example
-    /// ```sql
-    /// SELECT lpad('hello', 10, ' ');
-    /// ```
-    Lpad,
-    /// SQL 'repeat()' function
-    /// # Example
-    /// ```sql
-    /// SELECT repeat('X', 2);
-    /// ```
-    Repeat,
-    // Like,
-    // Ilike,
-    /// SQL 'substring()' function
-    /// # Example
-    /// ```sql
-    /// SELECT substring('hello', 1, 2);
-    /// ```
-    Substr,
-    /// SQL 'to_date()` function
-    /// # Example
-    /// ```sql
-    /// SELECT to_date('2021-01-01', 'YYYY-MM-DD');
-    /// ```
-    ToDate,
-    /// SQL 'to_datetime()' function
-    /// # Example
-    /// ```sql
-    /// SELECT to_datetime('2021-01-01 00:00:00', 'YYYY-MM-DD HH:MM:SS');
-    /// ```
-    ToDatetime,
+/// [SQLFunction] extends [FunctionExpr] with basic input validation (arity-check).
+pub struct SQLFunction(FunctionExpr);
```

Comment:
the issue with using `FunctionExpr` here is that it prevents us from using our newer exprs that have been moved over to `daft-functions`. Eventually we plan to migrate all of the `FunctionExpr` impls over to the newer `ScalarUDF` instead.

Reply:
To make this work with aggregates (`max`), as well as `ScalarUDF`'s, we'll likely want to use a trait object instead

```rs
pub trait SQLFunction {/*...*/}

pub struct SQLFunctions {
    map: HashMap<String, Arc<dyn SQLFunction>>,
}
```